### PR TITLE
Support for construction interception

### PR DESCRIPTION
**Pull request by Davidhanson90**
This PR adds base level support for interceptions.  Primarily construction interceptors.  This supports scenarios where developers require two stage initialization.  Currently not supporting sub classing.  I can do this later but didn't seem super urgent. 

All interceptors are registered in the root and will execute for for target types constructed in scopes.  See readme for usage notes.

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -677,6 +677,45 @@ level1.isDestroyed() //True;
...
+
+* `target` -  The type we wish to intercept
+* `beforeCreate` - A method that will be invoked directly before the type instanced and after all its constructor args are resolved. 
+* `afterCreate` -  A method that will be invoked immediately after the target type was instanced.  
+
+Each method call will receive an `injection context` object which provides information about the context of injection. This includes information such as the injector instance resolving the type, the configuration used during construction and an array of constructor args.
+
+## Registering an interceptor
+
+The easiest way to register an interceptor is to simply decorate your class with `@Interceptor()`.  This will automatically register your interceptor with the root injector.  However if you want to use the injector API you can also do as follows.
```
Comment (Roaders) [inline readme.md:711]: Is this enforced? Can you decorate a class that does not implement `IConstructionInterceptor` with `@Interceptor()`?

File: main/contracts/contracts.ts [hunk 2]
```diff
@@ -409,3 +421,49 @@ export interface IMetricsProvider extends IMetrics {
...
+
+/**
+ * Base constructor injector interceptor interface
+ * @description Constructor Interceptors only fire when cache missed
+ */
+export interface IConstructionInterceptor<TTarget extends Newable = any> {
+    /**
+     * The type targeted for interception
+     */
+    target: TTarget;
```
Comment (Roaders) [inline main/contracts/contracts.ts:455]: this should probably be `readonly`

**Review by Roaders — CHANGES_REQUESTED** [review]

File: readme.md (hunk 1 above)
Comment (Davidhanson90) [reply readme.md:711]: When I try and constrain this I get complaints about the type not implementing the methods.  See below.  Removing the default generic value makes no diff.  

image

Comment (Roaders) [reply readme.md:711]: I think that I have tried to enforce what types can be decorated before and it's not that straightforward. I'll have a go in playground to see if I can get it to work.

Comment (Roaders) [reply readme.md:711]: This seems to work:

```ts
interface IMyInterface{
    interestedIn: (constructor: new (...args: any[]) => any) => boolean;
}


export function Interceptor() {
    return (interceptor: new (...args: any[]) => IMyInterface) => {
    };
}

@Interceptor() // decoration works
class InterfaceImplementation implements IMyInterface{
    interestedIn(constructor: new (...args: any[]) => any): boolean {
        return false
    }
}

@Interceptor() // decoration does not work
class NonInterfaceImplementation{
}
```

[playground](https://www.typescriptlang.org/play?ts=3.8.3#code/FASwdgLgpgTgZgQwMZQAQEkCyBPdlaIoDewqZq40MUAztACZ4BcqAFEgPZh0wCuSEDjBZgoAdzYA6aQhgBzGiwRhsAbQC6ASlQBeAHypl2bftQAjDhwA2UZQG5gAX2AuoADwAOQiKji8wAiBcGPgwKB6CMKzaJOSo1BC8MGBslLDhkSLiUjLyioYqGiYGWLihhFDFqLHkjg7OwAACeFQZQtGoAPSdqPRQnDAIEEEpYkIA1jTASFYINDQhVBXoALYeNitQkEMjFGsbWxALpS0EyFA1ZGnUdFCMYOxcPPyZqKISrNKSsgpKhVq6AxGTQsCzWWwpS5xeJQRLJXwIKw0KCkWpOFzNUJtKLabq9fpCHbBegcWhvDg+MYwcbTWbzVAAOS4p3g51W6ygm22wy4JGcQA)

Comment (Davidhanson90) [reply readme.md:711]: This is not using generic params and then applying those to methods.  Not equal basically to what we have in the PR. Try here https://www.typescriptlang.org/play?ts=3.5.1#code/KYDwDg9gTgLgBDAnmYcBywDuBDARgG1QF44A7LOACgDpbsoBzAZwC45tTEBtAXQEo4RAHztOAbgBQE0JFhwAZgFdSAYxgBLCKTgBJUjGBQVwMDGiUBAbwlxbcKMBiKo2yuv2Hjp6Gx0BhLSYYKEU1TVI9AyMTMygBYThrOzsbOwBfSTSpaXBoeHco+WxjXT0AK2AwrQCPEBgAHgAVRuRUUANSABMmdCw8QkFRRBEkuwB6ACoJ1LsJuEaAC1QkFDhcYHcGOBVA4NCDTpnbCbGjh2xOrXxEBFa2ZtbJZMnp5Lg5xdR3CrVoBAX1D13EEOCVAewmEwICp1NgDggIBAju9TslzpdSNc4N9KrE2BxEE9xlNkR8ltstPJ1AxnHDwnBFExgJ04J1nJsKaQgiEqqRSai7OirjcdqQqTSoHStPjxEcXqS4DVufs-vQaQBbYD6HqM5kKP6i5W8-lnYAXYWco2xACCjFYit2PNiAAV6NhNVEmE0WighJkpPL3nAAELYJmWva-KDY0g-WIxqJeeMFQxFYw2OYAAU6wCYKig6lM9KVkfjkU8MWgPQtVIccEwS20KmK5PVgKZhxROVk+Q8UDTqH8jv24XL0W8UG9asccHaWu6vRwBGIQxGcpJbzJy1aCGn8PkfxT440WhNaLNGKxMD390ae8k69eyTmegAbhAANZ69YHuswckrF8QJcteqjfoo8DYPIUTYjAPSGqWqqMIomranACzYK+qDrFq9i5hA+BYZ2z5wJmYBuuqnIGHUDq1JB8CQAUcAQPIEZOuEZ52D+0DAH45wGJQorUTAvjlLi4Q1MJU6MI4Qh8Gw77qJ0D7PBuJFvp+eqdOoDhqFiUEwf+26rBhPQ4dowKgcYxGzKR5GSpRlmgqgnwxiCYHMaxRlwAw6hYdogEKmRFFUe0tHCewDEQExLFsSOp6bgKtgGYYfFmgJTlgb4IHOT6wDSQwskADShXUomxuJ1RaFJzR7nJCnRcpEhZBIKj4GGPQAKKkL55CJC1mZjkm5h8N2eTbO1kJwN1vXAENlbRuo6pgIQqFwboJbsVo80TvUgGxTN7jAGuyRgIoBDqCoazAL+c2kO+zYntoJDkJgcC2pKiD1DoYm8pJ7R7a0B09UdQhCBYRK2GdF1XSlUAaY99IvRQH3YF9lhuVZwBsId5BiKVImlBVf3VQD+2sbjx1wGk4N8CpdjQ-gl27jJhPk9NIN9SQlOQ3AjPM9xDhpXCwCCaTZVE3GEniw07OU-VcCKSyoxvP8gLUILd0PVKpDUGdTALGLdF00cWSnedTOw9BqX8aLmXGDjnPACVQntOVUtVXRgMoMDs0K0riTInY-7q3DCM63rjKGxj9vOwT1Mm8kWRZEAA

Comment (Davidhanson90) [reply readme.md:711]: Ok I need to rework this then.  As I need an instance to be registered.

Comment (Davidhanson90) [reply readme.md]: Ok fixes in and docs updated

**Review by Roaders — APPROVED** [review]

**Pull request merged**